File: src/main/java/insogroup/HtmlParser/StandardRealisation/PluginSyntaxAnalyzer/StandardState.java
package insogroup.HtmlParser.StandardRealisation.PluginSyntaxAnalyzer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import insogroup.HtmlParser.StandardRealisation.PluginSyntaxAnalyzer.interfaces.State;
import insogroup.HtmlParser.StandardRealisation.exceptions.PluginException;
import insogroup.HtmlParser.StandardRealisation.settings.interfaces.Settings;
import org.xml.sax.Attributes;


public class StandardState implements State {
    protected State parent;
    protected List<State> childs;

    protected Settings settings;

    protected String tag;
    protected String text;

    protected HashMap<String, String> attributes;

    public StandardState(State parent, String tag, Attributes attributes, Settings settings) {
        this.parent = parent;
        this.tag = tag;
        this.settings = settings;
        setAttributes(attributes);
        this.childs = new ArrayList<>();
    }

    protected void setAttributes(Attributes attributes){
        this.attributes = new HashMap<>();
        for (int i = 0; i < attributes.getLength(); i++){
            String key = attributes.getQName(i);
            String value = attributes.getValue(i);
            this.attributes.put(key, value);
        }
    }

    public State addChild(String name, Attributes attr){
        State childState = new StandardState(this, name, attr, settings);
        childs.add(childState);
        return childState;
    }

    public String get(String key) throws PluginException{
        try{
            return attributes.get(key);
        }
        catch (Exception e){
            throw new PluginException("Can't give plugin attribute: " + key + ": " + e.getMessage());
        }
    }

    public State getParent() throws PluginException{
        if (parent == null){
            throw new PluginException("Is root");
        }
        return parent;
    }

    public String getTag() {
        return tag;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        String res = "";
        for (char tmp : text.toCharArray()){
            if (!settings.checkStandardRegexp("" + tmp)){
                continue;
            }
            res += tmp;
        }
        this.text = res;
    }

    public List<State> getChilds() {
        return childs;
    }

    public State getRoot(){
        State res = this;
        try{
            while (true){
                res = res.getParent();
            }
        }
        catch (PluginException e){
            return res;
        }
    }

    public HashMap<String, String> getAttributes() {
        return attributes;
    }



}

File: src/main/kotlin/com/github/insanusmokrassar/HTMLParser/StandardRealisation/PluginSyntaxAnalyzer/StandardState.kt
package com.github.insanusmokrassar.HTMLParser.StandardRealisation.PluginSyntaxAnalyzer

import java.util.ArrayList

import com.github.insanusmokrassar.HTMLParser.StandardRealisation.PluginSyntaxAnalyzer.interfaces.State
import com.github.insanusmokrassar.HTMLParser.StandardRealisation.exceptions.PluginException
import com.github.insanusmokrassar.HTMLParser.StandardRealisation.Settings
import com.github.insanusmokrassar.HTMLParser.StandardRealisation.checkStandardRegexp
import org.xml.sax.Attributes

open class StandardState(
        override val parent: State?,
        override val tag: String,
        externalAttributes: Attributes,
        protected val settings: Settings
) : State {
    override var text: String = ""
        set(value) {
            var res = ""
            value.toCharArray().forEach {
                if (settings.checkStandardRegexp("" + it)) {
                    res += it
                }
            }
            field = res
        }
    override val childs: MutableList<State> = ArrayList()

    override val attributes: Map<String, String> = (0 until externalAttributes.length).map {
        i ->
        Pair(externalAttributes.getQName(i), externalAttributes.getValue(i))
    }.toMap()

    override val root: State
        get() {
            var result: State = this
            while (true) {
                result = result.parent ?: return result
            }
        }

    override fun addChild(name: String, attr: Attributes): State {
        return StandardState(this, name, attr, settings).also {
            childs.add(it)
        }
    }

    override fun get(key: String): String {
        return attributes[key] ?: throw PluginException("Can't give plugin attribute: $key")
    }
}
